Mage.Sets/src/mage/cards/o/OsgirTheReconstructor.java

```diff
@@ -0,0 +1,131 @@
+
+        // Vigilance
+        this.addAbility(VigilanceAbility.getInstance());
+
+        // {1}, Sacrifice an artifact: Target creature you control gets +2/+0 until end of turn.
+        Ability ability = new SimpleActivatedAbility(Zone.BATTLEFIELD, new BoostTargetEffect(2, 0, Duration.EndOfTurn), new ManaCostsImpl<>("{1}"));
+        ability.addTarget(new TargetCreaturePermanent());
+        ability.addCost(new SacrificeTargetCost(new TargetControlledPermanent(new FilterControlledArtifactPermanent("an artifact"))));
+        this.addAbility(ability);
+
+        // {X},{T}, Exile an artifact with mana value X from your graveyard: Create two tokens that are copies of the exiled card. Activate only as
+        Ability copyAbility = new ActivateAsSorceryActivatedAbility(Zone.BATTLEFIELD,
+                new OsgirTheReconstructorCreateArtifactTokensEffect(),
+                new ManaCostsImpl("{X}"));
+        copyAbility.addCost(new TapSourceCost());
+        copyAbility.setTargetAdjuster(OsgirTheReconstructorAdjuster.instance);
+
+        this.addAbility(copyAbility);
+    }
+
+    private OsgirTheReconstructor(final OsgirTheReconstructor card) {
+        super(card);
+    }
+
+    @Override
+    public OsgirTheReconstructor copy() {
+        return new OsgirTheReconstructor(this);
+    }
+}
+
+enum OsgirTheReconstructorAdjuster implements TargetAdjuster {
+    instance;
+
+    @Override
+    public void adjustTargets(Ability ability, Game game) {
+        int xValue = ability.getManaCostsToPay().getX();
+        FilterCard filter = new FilterArtifactCard("an artifact card with mana value "+xValue+" from your graveyard");
+        filter.add(new ManaValuePredicate(ComparisonType.EQUAL_TO, xValue));
+        ability.getTargets().clear();
+        ability.getTargets().add(new TargetCardInYourGraveyard(filter));
+    }
+}
+
+class OsgirTheReconstructorCreateArtifactTokensEffect extends OneShotEffect {
+
+    public OsgirTheReconstructorCreateArtifactTokensEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Create two tokens that are copies of the exiled card.";
+    }
+
+    public OsgirTheReconstructorCreateArtifactTokensEffect(final OsgirTheReconstructorCreateArtifactTokensEffect effect)  {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player player = game.getPlayer(source.getControllerId());
+        Card card = game.getCard(targetPointer.getFirst(game, source));
+
+        player.moveCards(card, Zone.EXILED, source, game);
+
+        if (player == null || card == null) {
+            return false;
+        }
+
+        CreateTokenCopyTargetEffect effect = new CreateTokenCopyTargetEffect(player.getId(), null, false, 2);
+        effect.setTargetPointer(new FixedTarget(card.getId(), game.getState().getZoneChangeCounter(card.getId())));
+        effect.apply(game, source);
+
+        return  true;
+    }
+
+    @Override
+    public Effect copy() {
```

Comment:
this should return `OsgirTheReconstructorCreateArtifactTokensEffect`

Reply:
this does return that no?

Reply:
I mean that it should be `public OsgirTheReconstructorCreateArtifactTokensEffect copy() {`, but I'm not sure if it even matters

Reply:
got it